src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -377,6 +387,117 @@ public void WriteValue(string value)
+                }
+                else
+                {
+                    WriteStringChunk(chunk, isFinalBlock);
+                }
+
+                // Flush the buffer if needed
+                this.FlushIfBufferThresholdReached();
+            }
+
+            this.bufferWriter.Write(doubleQuote.Slice(0, 1).Span);
+
+            // since we bypass the Utf8JsonWriter, we need to signal to other
+            // Write methods that a separator should be written first
+            if (this.isWritingAtStartOfArray || this.isWritingConsecutiveRawValuesAtStartOfArray || !this.IsInArray())
+            {
+                this.shouldWriteSeparator = true;
+            }
+        }
+
+        private  unsafe int NeedsEscaping(ReadOnlySpan<char> value)
+        {
+            // Some implementations of JavaScriptEncoder.FindFirstCharacterToEncode may not accept
+            // null pointers and guard against that. Hence, check up-front to return -1.
+            if (value.IsEmpty)
+            {
+                return -1;
+            }
+
+            fixed (char* ptr = value)
+            {
+                return this.encoder.FindFirstCharacterToEncode(ptr, value.Length);
+            }
+        }
+
+        private void WriteEscapedStringChunk(ReadOnlySpan<char> sourceChunk, int firstIndexToEscape, bool isFinalBlock)
+        {
+            char[] valueArray = null;
+            const int StackallocCharThreshold = 128;
+
+            int maxSize = (firstIndexToEscape + 6) * (sourceChunk.Length * firstIndexToEscape);
+
+            Span<char> destination = maxSize <= StackallocCharThreshold
+                ? stackalloc char[maxSize]
+                :valueArray = ArrayPool<char>.Shared.Rent(maxSize);
+
+            try
+            {
+                OperationStatus status = this.encoder.Encode(
+                    sourceChunk,
+                    destination,
+                    out int bytesConsumed,
+                    out int bytesWritten,
+                    isFinalBlock
+                );
+
+                if (status == OperationStatus.Done)
+                {
+                    WriteStringChunk(destination.Slice(0, bytesWritten), isFinalBlock);
+                }
+                else
+                {
+                    Console.WriteLine($"Encoding failed with status: {status}");
+                }
+            }
+            finally
+            {
+                if (valueArray != null)
+                {
+                    ArrayPool<char>.Shared.Return(destination.ToArray());
+                }
+            }
+        }
+
+        private void WriteStringChunk(ReadOnlySpan<char> chunk, bool isFinalBlock)
+        {
+            int maxUtf8Length = Encoding.UTF8.GetMaxByteCount(chunk.Length);
+
+            // technically this might resize the buffer. But if we use a small chunk size, say 4K, and since we're flushing
+            // after every chunk, this is unlikely even if we consider the max byte count

```

Comment:
I think since we're not buffering after every chunk, and we're using a 2k chunk size, you should update or remove this comment.